### PR TITLE
feat: added typescript testcaferc configuration file support

### PR DESCRIPTION
**Pull request by Bayheck**
## Purpose
_Describe the problem you want to address or the feature you want to implement._
Added support for TypeScript configuration file (testcaferc.ts) so TS code can be referenced in global hooks and similar functionality. However, defining options for testcaferc.ts compilation itself is only possible from CLI. We can define them using 
`--compiler-options`. If no options are provided in CLI  we use default compiler with default compilation options.

## Approach
_Describe how your changes address the issue or implement the desired functionality in as much detail as possible._
Added new methods in \src\configuration\testcafe-configuration.ts:  _readTsConfigurationFileContent, _isTSConfiguration and _readConfigurationOptions.

Added new method in \src\compiler\test-file\formats\typescript\compiler.ts: compileConfiguration.

Inside testcafe-configuration.ts If config file is a ts config, then call _readTsConfigurationFileContent.

Also in \src\index.js pass compilerOptions to configuration.init().

Wrote tests for testing the above mentioned functionality inside \test\server\configuration-test.js

_Provide a link to the existing issue(s), if any._
closes https://github.com/DevExpress/testcafe/issues/7373

## Pre-Merge TODO
- [x] Write tests for your proposed changes
- [x] Make sure that existing tests do not fail

### CODE REVIEW

File: src/compiler/test-file/formats/typescript/compiler.ts [hunk 1]
```diff
@@ -267,4 +267,22 @@ export default class TypeScriptTestFileCompiler extends APIBasedTestFileCompiler
...
+        const [compiledCode] = await this.precompile([{ code: '', filename }]);
+
+        if (compiledCode) {
+            this._setupRequireHook({ });
+
+            compiledConfigurationModule = await this._runCompiledCode({}, compiledCode, filename);
+
+            this._removeRequireHook();
+
+            return compiledConfigurationModule.exports;
```
[the remark this answers is not in the document]
Comment (aleks-pro) [reply src/compiler/test-file/formats/typescript/compiler.ts]: I can see that you added check here but why `_execAsModule` method even can return null for case when we compile a config file?

**Review by aleks-pro — COMMENTED** [review]
https://github.com/DevExpress/testcafe/pull/8039#discussion_r1406087292

File: src/compiler/test-file/formats/typescript/compiler.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Bayheck) [reply src/compiler/test-file/formats/typescript/compiler.ts]: Because if this.esm inside of _execAsModule is true then _execAsModule returns undefined.

**Review by aleks-pro — APPROVED** [review]

**Comment by lukas-zech-software** [comment]
is there any progress here? 

We could really use this feature and everything in this PR seem Green and Approved.
Anything missing?

**Comment by Bayheck** [comment]
@lukas-zech-software
We are not merging this PR since we think this feature still needs improvement. I cannot give you an ETA at this time since my predictions may not be accurate.

**Comment by Jackman3005** [comment]
So excited for this!

**Pull request closed without merging**